rfcs/text/0015_ols_phase_1.md

```diff
@@ -0,0 +1,260 @@
+
+See [unresolved question 3](#83-behavior-when-security-is-disabled)
+
+# 4. Drawbacks
+
+As outlined above, this approach introduces additional overhead to many of the saved object APIs. We minimize this by denoting which saved object types require this additional authorization.
+
+This first phase also does not allow a public object to become private. Search sessions may migrate to OLS in the future, but this will likely be a coordinated effort with Elasticsearch, due to the differing ownership models between OLS and async searches.
+
+# 5. Alternatives
+
+OLS can be thought of as a Kibana-specific implementation of [Document level security](https://www.elastic.co/guide/en/elasticsearch/reference/current/document-level-security.html) ("DLS"). As such, we could consider enhancing the existing DLS feature to fit our needs. This would involve considerable work from the Elasticsearch security team before we could consider this, and may not scale to subsequent phases of OLS.
+
+# 6. Adoption strategy
+
+Adoption for net-new features is hopefully straightforward. Like most saved object features, the saved objects service will transparently handle all authorization and auditing of these objects, so long as they are properly registered.
+
+Adoption for existing features (public saved object types) is not addressed in this first phase.
+
+# 7. How we teach this
+
+Updates to the saved object service's documentation to describe the different `classification`s would be required. Like other saved object security controls, we want to ensure that engineers understand that this only "works" when the security wrapper is applied. Creating a bespoke instance of the saved objects client, or using the raw repository will intentionally bypass these authorization checks.
+
+# 8. Unresolved questions
+
+## 8.1 `acl.owner`
+
+The `acl.owner` property will uniquely identify the owner of each `private` saved object. We are still iterating with the Elasticsearch security team on what this value will ultimately look like. It is highly likely that this will not be a human-readable piece of text, but rather a GUID-style identifier.
+
+## 8.2 Authorization for private objects
+
+The user identified by `acl.owner` will be authorized for all operations against that instance, provided they pass the existing type/space/action authorization checks.
+
+In addition to the object owner, we also need to allow administrators to manage these saved objects. This is beneficial if they need to perform a bulk import/export of private objects, or if they wish to remove private objects from users that no longer exist. The open question is: **who counts as an administrator?**
+
+We have historically used the `Saved Objects Management` feature for these administrative tasks. This feature grants access to all saved objects, even if you're not authorized to access the "owning" application. Do we consider this privilege sufficient to see and potentially manipulate private saved objects?
```

Comment:
I've been asking myself this question as I read this RFC. I'm not sure what I think the right answer is yet, I need to think about it some more. A stricter (safer/better?) alternative may be simply anyone who has the `kibana_admin` role. I'll follow up on this in another comment.

Reply:
Yeah that was my other thought as well. A third idea is to introduce a sub-feature privilege to Saved Objects Management to allow administrators to control this access in a more granular manner

Reply:
> A third idea is to introduce a sub-feature privilege to Saved Objects Management to allow administrators to control this access in a more granular manner

I like this idea. Having "all" access to saved-object management is already a rather elevated set of privileges. Extending it to also include the ability to create private objects for others seems inline with the existing set of privileges, but I could see users wanting to revoke this ability.

Reply:
+1!

Reply:
Alright, sounds like we're all in agreement for introducing a sub-feature privilege. I'll update the RFC to capture this.